### PR TITLE
Tests folder layout

### LINKED ISSUE #270: Restructure `tests/` directory

### PR DESCRIPTION
**Pull request by 5K1PY**
Fixes #270

### CODE REVIEW

File: pisek/task_jobs/data/testcase_info.py [hunk 1]
```diff
@@ -48,11 +48,31 @@ def mixed(name: str) -> "TestcaseInfo":
     def static(name: str) -> "TestcaseInfo":
         return TestcaseInfo(name, 1, TestcaseGenerationMode.static, False)
 
-    def input_path(self, env: Env, seed: Optional[int] = None) -> TaskPath:
+    def input_path(
+        self, env: Env, seed: Optional[int] = None, solution: Optional[str] = None
```
Comment (SvizelPritula) [inline pisek/task_jobs/data/testcase_info.py:52]: What is the `solution` parameter for?

File: pisek/task_jobs/data/testcase_info.py [hunk 2]
```diff
@@ -48,11 +48,31 @@ def mixed(name: str) -> "TestcaseInfo":
...
+    def reference_output(
+        self, env: Env, seed: Optional[int] = None, solution: Optional[str] = None
+    ) -> OutputPath:
+        is_static = self.generation_mode == TestcaseGenerationMode.static
+
+        input_path = self.input_path(env, seed, solution=env.config.primary_solution)
+        if is_static:
+            path = OutputPath.static(input_path.replace_suffix(".out").name)
+        else:
+            path = input_path.to_output()
```
Comment (SvizelPritula) [inline pisek/task_jobs/data/testcase_info.py:71]: What's different about static outputs?

File: pisek/utils/paths.py [hunk 3]
```diff
@@ -106,74 +113,69 @@ def data_path(env: "Env", *path: str) -> "TaskPath":
...
 
-    @staticmethod
-    def invalid_file(env: "Env", name: str, seed: int) -> "TaskPath":
-        name = os.path.splitext(name)[0]
-        return TaskPath.invalid_path(env, f"{name}.{seed:x}.invalid")
 
-    @staticmethod
-    def output_path(env: "Env", *path: str) -> "TaskPath":
-        return TaskPath.data_path(env, OUTPUTS_SUBDIR, *path)
+class SanitizeablePath(TaskPath):
```
Comment (SvizelPritula) [inline pisek/utils/paths.py:124]: Sanitizeable is not a real word, Sanitizable is.

**Review by SvizelPritula — CHANGES_REQUESTED** [review]

**Comment by SvizelPritula** [comment]
It might be a good idea to prevent users from making a solution called `_inputs`, for example, by prohibiting solution names from statring with `_`. Although, I don't expect anyone to make such a solution.

File: pisek/task_jobs/data/testcase_info.py (hunk 2 above)
Comment (5K1PY) [reply pisek/task_jobs/data/testcase_info.py:71]: Refernce static outputs are static, where others are generated by primary solution.

File: pisek/task_jobs/data/testcase_info.py (hunk 1 above)
Comment (5K1PY) [reply pisek/task_jobs/data/testcase_info.py:52]: To getting input in directory corresponding to right solution

File: pisek/task_jobs/data/testcase_info.py (hunk 2 above)
Comment (SvizelPritula) [reply pisek/task_jobs/data/testcase_info.py:71]: Does this function return the path for a solution-specific copy of the reference output, or to a single shared instance of the reference output?

Comment (5K1PY) [reply pisek/task_jobs/data/testcase_info.py:71]: Specific solution copy if `solution` is not `None`, otherwise `_inputs` instance.

Comment (SvizelPritula) [reply pisek/task_jobs/data/testcase_info.py:71]: I personaly found this functions logic to be hard to follow, but I think I understand it now.

File: pisek/config/task_config.py [hunk 4]
```diff
@@ -494,7 +499,13 @@ def convert_yesno(cls, value: str, info: ValidationInfo) -> bool:
...
     def convert_auto(cls, value: str, info: ValidationInfo) -> str:
         if value == "@auto":
-            return info.data.get("name", "")
+            value = info.data.get("name", "")
+
+        if value.startswith("_"):
+            raise PydanticCustomError(
+                "invalid_solution_source",
+                f"Solution source must not start with '_'",
+            )
```
Comment (SvizelPritula) [inline pisek/config/task_config.py:508]: Is there a reason for this?

Comment (5K1PY) [reply pisek/config/task_config.py:508]: What in particular you are objecting?

Comment (SvizelPritula) [reply pisek/config/task_config.py:508]: Names cannot start with a `_` to prevent collisions with reserved folder names, but I don't see a reason for the source file name not to be able to strart with `_`.

Comment (5K1PY) [reply pisek/config/task_config.py:508]: We name folders according to sources, not names. As to why enforce same thing is enforced on names, is that it otherwise produces confusing error messages:
```
In section [all_solutions], key 'source' (in config file /home/skipy/dev/pisek/fixtures/sum_cms/cms-defaults):
  Solution source must not start with '_': '@auto'
  ```

Comment (SvizelPritula) [reply pisek/config/task_config.py:508]: True, I forgot about that.
Why are we naming the folders according to the source file, anyway?

**Review by SvizelPritula — APPROVED** [review]
LGTM

File: pisek/config/task_config.py (hunk 4 above)
Comment (5K1PY) [reply pisek/config/task_config.py:508]: > Why are we naming the folders according to the source file, anyway?

Uups, that probably doesn't work once we allow for same source be different solutions differentiated by arguments. Gonna rewrite that.

**Review by SvizelPritula — APPROVED** [review]
LGTM
